# Week 2 — Credit Card Fraud Detection Pipeline
## Preprocessing, ADASYN and Model Training

**Business:** Select the best model for real-time fraud scoring.
We compare 4 models and select by Recall — missing fraud costs money.

**Dataset:** 284,807 transactions — 492 fraud — 0.17% fraud rate

**Goal:** Balance classes with ADASYN — train 4 models — compare — save best pipeline

**Author:** Martin James | github.com/M20Jay

## Section 1 - Import Libraries

In [ ]:
# ── Importing Libraries ───────────────────────────────────────
import pandas as pd
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, roc_auc_score, average_precision_score
from imblearn.over_sampling import ADASYN
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
import warnings
warnings.filterwarnings('ignore')

print("Libraries successfully imported ✅")

Libraries successfully imported ✅


## Section 2 - Load Data from PostgreSQL

Data lives in PostgreSQL — not in a CSV file. This is the production approach — any service can query the same database.  
The notebook, FastAPI, and Kafka all read from one single source of truth.

In [5]:
from sqlalchemy import create_engine, text
import pandas as pd
import os

# ── Database Connection ───────────────────────────────────────
DATABASE_URL = os.environ.get(
    "DATABASE_URL",
    "postgresql+psycopg2://martin:martin123@127.0.0.1:5432/fraud_db"
)

engine = create_engine(DATABASE_URL)
# ── Read from PostgreSQL ──────────────────────────────────────
print("Reading from PostgreSQL fraud_raw table...")

query = "SELECT * FROM fraud_raw"
df = pd.read_sql(query, engine)

print(f"Shape        : {df.shape}")
print(f"Fraud cases  : {(df['class']==1).sum():,}")
print(f"Legit cases  : {(df['class']==0).sum():,}")
print(f"Columns      : {list(df.columns)}")
print(f"\n✅ Data loaded from PostgreSQL successfully")

Reading from PostgreSQL fraud_raw table...
Shape        : (284807, 31)
Fraud cases  : 492
Legit cases  : 284,315
Columns      : ['time', 'v1', 'v2', 'v3', 'v4', 'v5', 'v6', 'v7', 'v8', 'v9', 'v10', 'v11', 'v12', 'v13', 'v14', 'v15', 'v16', 'v17', 'v18', 'v19', 'v20', 'v21', 'v22', 'v23', 'v24', 'v25', 'v26', 'v27', 'v28', 'amount', 'class']

✅ Data loaded from PostgreSQL successfully


## Section 3 - Feature Engineering & Train/Test Split

Amount and Time are scaled using StandardScaler.  
V1 to V28 are already PCA-scaled by Kaggle — left untouched.  
Split is performed BEFORE ADASYN — test set must never see synthetic samples.  
Violating this rule produces fake accuracy — a critical MLOps principle.

In [9]:
# ── Scale Amount and Time ─────────────────────────────────────
scaler = StandardScaler()
 
df['amount_scaled'] = scaler.fit_transform(df[['amount']])
df['time_scaled'] = scaler.fit_transform(df[['time']])

df_model = df.drop(columns =['amount', 'time'])

# ── Define Features and Target ────────────────────────────────
X = df_model.drop('class', axis=1)
y = df_model['class']

# ── Train/Test Split ──────────────────────────────────────────
X_train,X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42,stratify=y)

print(f"Train size     : {X_train.shape[0]:,}")
print(f"Test size      : {X_test.shape[0]:,}")
print(f"Features       : {X_train.shape[1]}")
print(f"Fraud in train : {y_train.sum():,}  ({y_train.mean()*100:.3f}%)")
print(f"Fraud in test  : {y_test.sum():,}  ({y_test.mean()*100:.3f}%)")
print(f"\n✅ Section 3 complete — data ready for ADASYN")

Train size     : 227,845
Test size      : 56,962
Features       : 30
Fraud in train : 394  (0.173%)
Fraud in test  : 98  (0.172%)

✅ Section 3 complete — data ready for ADASYN


## Section 4 - ADASYN Class Balancing

Fraud cases represent only 0.173% of all transactions — extreme imbalance.  
ADASYN generates synthetic fraud samples focused on the hardest cases to classify.  
Applied on training set ONLY — test set must remain untouched and real.  
Applying ADASYN on test data would give fake evaluation results.

In [21]:
# ── Apply ADASYN on Training Set Only ────────────────────────
print("Before ADASYN:")
print(f"  Fraud : {(y_train==1).sum():,}")
print(f"  Legit : {(y_train==0).sum():,}")
print(f"  Ratio : {y_train.mean()*100:.3f}%")

adasyn = ADASYN(random_state=42, n_neighbors=5)
X_train_bal, y_train_bal = adasyn.fit_resample(X_train, y_train)


print(f"\nAfter ADASYN:")
print(f"  Fraud : {(y_train_bal==1).sum():,}")
print(f"  Legit : {(y_train_bal==0).sum():,}")
print(f"  Ratio : {y_train_bal.mean()*100:.1f}%")
print(f"\n✅ Section 4 complete — balanced data ready for training")



Before ADASYN:
  Fraud : 394
  Legit : 227,451
  Ratio : 0.173%

After ADASYN:
  Fraud : 227,436
  Legit : 227,451
  Ratio : 50.0%

✅ Section 4 complete — balanced data ready for training


## Section 5 - Train and Compare 4 Models

Four models trained on ADASYN balanced data and evaluated on real test data.  
Primary metric is PR-AUC — not accuracy and not ROC-AUC.  
Accuracy is misleading on imbalanced data — a model predicting all legitimate would score 99.8% accuracy but catch zero fraud.  
PR-AUC measures performance on the minority class only — fraud detection specifically.

In [ ]:
# ── Train and Compare 4 Models ────────────────────────────────
model ={
    "Logistic Regression" : LogisticRegression(max_iter=1000, random_state=42),
    "Random Forest"       : RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
    "XGBoost"             : XGBClassifier(eval_metric='logloss', random_state=42, n_jobs=-1),
    "LightGBM"            : LGBMClassifier(random_state=42, n_jobs=-1, verbose=-1)
}

results = {}
for name, model in model.items():
    print(f"\nTraining : {name}")
    t0 = time.time()

model.fit(X_train_bal, y_train_bal)
y_pred = model.redict(X_test)
y_pred_proba = model.predict_proba(X_test)[:,1]

roc   = roc_auc_score(y_test, y_proba)


SyntaxError: incomplete input (3706016692.py, line 10)